bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+
+    public Map<String, TransactionLogRecord> getPendingLogs() {
+        Map<String, TransactionLogRecord> pendingTransactions = new HashMap<>();
+        Map<String, TransactionLogRecord> transactionLogs = readLogsFromFile(file);
+        if (transactionLogs == null) {
+            return null;
+        }
+        for (Map.Entry<String, TransactionLogRecord> entry : transactionLogs.entrySet()) {
+            String trxId = entry.getKey();
+            TransactionLogRecord trxRecord = entry.getValue();
+
+            if (!trxRecord.isCompleted()) {
+                pendingTransactions.put(trxId, trxRecord);
+            }
+        }
+        return pendingTransactions;
+    }
+
+    /**
+     * Write a transaction log entry to the recovery log file.
+     *
+     * @param str the log entry to write
+     */
+    private void writeToFile(String str, boolean force) {
+        if (appendChannel == null || !appendChannel.isOpen()) {
+            initAppendChannel(file);
+        }
+        byte[] bytes = str.getBytes();
+        try {
+            appendChannel.write(java.nio.ByteBuffer.wrap(bytes));
+            appendChannel.force(force);
+        } catch (IOException e) {
+            stderr.println("error: failed to write to recovery log file " + file.toPath());
+        }
+    }
+
+
+    private Map<String, TransactionLogRecord> readLogsFromFile(File file) {
+        Map<String, TransactionLogRecord> logMap = new HashMap<>();
+        if (!file.exists() || file.length() == 0) {
+            return null;
+        }
+        if (appendChannel != null) {
+            closeEverything();
+        }
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                TransactionLogRecord transactionLogRecord = TransactionLogRecord.parseTransactionLogRecord(line);
+                if (transactionLogRecord != null) {
+                    // TODO: add a check here to check whether the record exists in the logMap and new state is a valid
+                    //  state from the current state
+                    logMap.put(transactionLogRecord.transactionId, transactionLogRecord);
+                }
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to read the recovery log file " + file.toPath());
+        }
+        return logMap;
+    }
+
+
+    private void cleanUpFinishedLogs() {
+        if (existingLogs.isEmpty()) {
+            return;
+        }
+        Iterator<Map.Entry<String, TransactionLogRecord>> iterator = existingLogs.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, TransactionLogRecord> entry = iterator.next();
+            if (entry.getValue().isCompleted()) {
+                iterator.remove(); // Safely remove the entry
+            }
+        }
+    }
+
+    public void ifNeedWriteCheckpoint() {
```

Comment:
Feel like better to rephrase it otherway around.
```suggestion
    public void writeCheckpointIfNeeded() {
```

Reply:
+1